gematria/datasets/process_and_filter_bbs_lib.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gematria/datasets/process_and_filter_bbs_lib.h"
+
+#include "X86InstrInfo.h"
+#include "gematria/llvm/disassembler.h"
+#include "gematria/llvm/llvm_architecture_support.h"
+#include "gematria/utils/string.h"
+#include "llvm/ADT/StringExtras.h"
+
+using namespace llvm;
+
+namespace gematria {
+
+BBProcessorFilter::BBProcessorFilter() {
+  LLVMSupport = LlvmArchitectureSupport::X86_64();
+  InstructionPrinter = LLVMSupport->CreateMCInstPrinter(0);
+}
+
+Expected<std::string> BBProcessorFilter::processBasicBlock(
+    const StringRef BasicBlock, const StringRef Filename,
+    bool FilterMemoryAccessingBlocks) {
+  // TODO(boomanaiden154): Update this to use llvm::Expected once
+  // gematria::ParseHex is refactored to return llvm::Expected.
+  auto MachineCodeHex = gematria::ParseHexString(BasicBlock);
+  if (!MachineCodeHex.has_value()) {
+    return createFileError(
+        Filename,
+        make_error<StringError>(
+            llvm::Twine("Could not parse: '").concat(BasicBlock).concat("'"),
+            std::make_error_code(std::errc::invalid_argument)));
+  }
+
+  Expected<std::vector<gematria::DisassembledInstruction>>
+      DisassembledInstructionsOrErr = gematria::DisassembleAllInstructions(
+          LLVMSupport->mc_disassembler(), LLVMSupport->mc_instr_info(),
+          LLVMSupport->mc_register_info(), LLVMSupport->mc_subtarget_info(),
+          *InstructionPrinter, 0, *MachineCodeHex);
+
+  if (!DisassembledInstructionsOrErr)
+    return createFileError(Filename, DisassembledInstructionsOrErr.takeError());
+
+  std::string OutputBlock;
+
+  for (const gematria::DisassembledInstruction &Instruction :
+       *DisassembledInstructionsOrErr) {
+    MCInstrDesc InstDesc =
+        LLVMSupport->mc_instr_info().get(Instruction.mc_inst.getOpcode());
+    // TODO(boomanaiden154): This filtering is a bit simplistic currently. We
+    // should probably be using MCInsrtDesc::hasUnmodeledSideEffects, but this
+    // needs to be evaluated at scale.
```

Comment:
In general, +1 for `hasUnmodeledSideEffects`, but at some point in the past IIRC it covered also instructions like `rep movs` that we might want to model.

Reply:
Right. Punting it off for now given discussion in previous patches.

This was just mechanically copied over from the executable.